File: C#/Leetcode/BinaryTree/LongestUnivaluePath.cs
using LeetcodeSolutions.DataStructures;

namespace LeetcodeSolutions.BinaryTree
{
    // Leetcode 687 - https://leetcode.com/problems/longest-univalue-path
    // Submission Detail - https://leetcode.com/submissions/detail/142000453/
    // Based on the intuition from LCD-543 Diameter of a Binary Tree

    public class LongestUnivaluePath
    {
        private int maxLength;

        // Tx = O(n)
        // Sx = O(d) for call stack

        public int FindLongestUnivaluePath(BinaryTreeNode root)
        {
            if (root != null) 
                FindLongestUnivaluePath(root, root.Val);

            return maxLength;
        }

        public int FindLongestUnivaluePath(BinaryTreeNode node, int parentValue)
        {
            if (node == null) return 0;

            int left = FindLongestUnivaluePath(node.Left,  node.Val);
            int right = FindLongestUnivaluePath(node.Right, node.Val);

            maxLength = System.Math.Max(left + right, maxLength);

            return node.Val == parentValue ?  1 + System.Math.Max(left, right): 0;
        }
    }
}

File: Java/DFS/Tree/LongestUnivaluePath.java
package DFS.Tree;

import DataStructures.TreeNode;

public class LongestUnivaluePath {
  // Leetcode 687: https://leetcode.com/problems/longest-univalue-path/description/
  // Submission Detail: https://leetcode.com/problems/longest-univalue-path/submissions/1650348537

  // Tx = O(n)
  // Sx = O(n)
  int result = 0;
  public int longestUnivaluePath(TreeNode root) {
    if(root == null)
      return 0;

    dfs(root, root.val);

    return result;
  }

  private int dfs(TreeNode node, int parentVal) {
    if(node == null)
      return 0;

    int left = dfs(node.left, node.val);
    int right = dfs(node.right, node.val);
    result = Math.max(result, left + right);

    if(node.val == parentVal)
      return 1 + Math.max(left, right);
    else
      return 0;
  }
}
